File: gameSrc/classes/uiButtonGroup.js
/**
 * @class
 */
cwt.UIButtonGroup = my.Class(/** @lends cwt.UIButtonGroup.prototype */ {

  constructor: function () {
    this.elements = [];
    this.selected = -1;
  },

  /**
   *
   * @param {cwt.UIField} el
   */
  addElement: function (el) {
    this.elements.push(el);
    if (this.selected === -1 && el.action) {
      this.elements[this.elements.length - 1].inFocus = true;
      this.selected = this.elements.length - 1;
    }
  },

  /**
   * Returns the current active button.
   *
   * @return {cwt.UIField}
   */
  activeButton: function () {
    return this.elements[this.selected];
  },

  /**
   * Returns a button by it's key.
   *
   * @param {String} key
   * @return {cwt.UIField}
   */
  getButtonByKey: function (key) {
    for (var i = 0, e = this.elements.length; i < e; i++) {
      if (this.elements[i].key === key) {
        return this.elements[i];
      }
    }
    return null;
  },

  /**
   *
   * @param {RegExp} reg
   */
  getButtonsByReg: function (reg) {
    var arr = [];

    for (var i = 0, e = this.elements.length; i < e; i++) {
      if (reg.test(this.elements[i].key)) {
        arr.push(this.elements[i]);
      }
    }

    return arr;
  },

  /**
   * Updates the index of the selected button in interconnection to a given position.
   *
   * @param {Number} x
   * @param {Number} y
   * @return {boolean} true, if the index was updated, else false
   */
  updateIndex: function (x, y) {
    for (var i = 0, e = this.elements.length; i < e; i++) {

      // inactive element
      if (!this.elements[i].action || this.elements[i].inactive) {
        continue;
      }

      if (this.elements[i].positionInButton(x, y)) {
        if (i === this.selected) {
          return false;
        }

        this.elements[this.selected].inFocus = false;
        this.selected = i;
        this.elements[this.selected].inFocus = true;

        return true;
      }
    }
    return false;
  },

  /**
   *
   * @param input
   * @return {boolean} true, if the index was updated, else false
   */
  handleInput: function (input) {
    if (cwt.DEBUG) cwt.assert(input !== null);

    if (typeof input!== "number") {
      input = input.key;
    }

    var res = true;
    this.elements[this.selected].inFocus = false;

    switch (input) {
      case cwt.Input.TYPE_UP:
      case cwt.Input.TYPE_LEFT:
        do {
          this.selected--;
          if (this.selected < 0) {
            this.selected = this.elements.length - 1;
          }
        } while (!this.elements[this.selected].action || this.elements[this.selected].inactive);
        break;

      case cwt.Input.TYPE_RIGHT:
      case cwt.Input.TYPE_DOWN:
        do {
          this.selected++;
          if (this.selected >= this.elements.length) {
            this.selected = 0;
          }
        } while (!this.elements[this.selected].action || this.elements[this.selected].inactive);
        break;

      default :
        res = false;
    }

    this.elements[this.selected].inFocus = true;

    return res;
  },

  /**
   * Draws the screen layout into a given context of a canvas object.
   *
   * @param {CanvasRenderingContext2D} ctx
   */
  draw: function (ctx) {
    for (var i = 0, e = this.elements.length; i < e; i++) {
      var el = this.elements[i];

      if (!el.inactive) {
        el.draw(ctx);
      }
    }
  }
});

File: gameSrc/net/wolfTec/states/widgets/UiButtonGroup.java
package net.wolfTec.states.widgets;

import net.wolfTec.Constants;
import net.wolfTec.input.InputData;
import net.wolfTec.utility.Debug;

import org.stjs.javascript.Array;
import org.stjs.javascript.JSCollections;
import org.stjs.javascript.RegExp;
import org.stjs.javascript.dom.canvas.CanvasRenderingContext2D;

public class UiButtonGroup {

    protected Array<UiField> elements;
    protected int selectedElement;

    public UiButtonGroup() {
        this.elements = JSCollections.$array();
        selectedElement = Constants.INACTIVE_ID;
    }

    public void addElement(UiField el) {
        this.elements.push(el);
        if (this.selectedElement == Constants.INACTIVE_ID && el.action != null) {
            this.elements.$get(this.elements.$length() - 1).inFocus = true;
            this.selectedElement = this.elements.$length() - 1;
        }
    }

    public UiField activeButton() {
        return this.elements.$get(selectedElement);
    }

    public UiField getButtonByKey(String key) {
        for (int i = 0, e = this.elements.$length(); i < e; i++) {
            if (this.elements.$get(i).key == key) {
                return this.elements.$get(i);
            }
        }
        return null;
    }


    //
    //
    // @param {RegExp} reg
    //
    public Array<UiField> getButtonsByReg(RegExp reg) {
        Array<UiField> arr = JSCollections.$array();

        for (int i = 0, e = this.elements.$length(); i < e; i++) {
            if (reg.test(this.elements.$get(i).key)) {
                arr.push(this.elements.$get(i));
            }
        }

        return arr;
    }

    //
    // Updates the index of the selected button in interconnection to a given position.
    //
    // @param {Number} x
    // @param {Number} y
    // @return {boolean} true, if the index was updated, else false
    //
    public boolean updateIndex(int x, int y) {
        for (int i = 0, e = this.elements.$length(); i < e; i++) {

            // inactive element
            if (this.elements.$get(i).action == null || this.elements.$get(i).inactive) {
                continue;
            }

            if (this.elements.$get(i).isPositionInElement(x, y)) {
                if (i == this.selectedElement) {
                    return false;
                }

                this.elements.$get(this.selectedElement).inFocus = false;
                this.selectedElement = i;
                this.elements.$get(this.selectedElement).inFocus = true;

                return true;
            }
        }
        return false;
    }

    public void setIndex(int index) {
        if (index < 0 && index >= this.elements.$length()) {
            Debug.logCritical(UiField.LOG_HEADER, "IllegalIndexArgument");
        }

        this.elements.$get(this.selectedElement).inFocus = false;
        this.selectedElement = index;
        this.elements.$get(this.selectedElement).inFocus = true;
    }

    public boolean isInactive() {
        return false;
    }

    //
    //
    // @param input
    // @return {boolean} true, if the index was updated, else false
    //
    public boolean handleInput(InputData inputData) {
        boolean res = true;
        this.elements.$get(selectedElement).inFocus = false;

        switch (inputData.key) {
            case UP:
            case LEFT:
                do {
                    this.selectedElement--;
                    if (this.selectedElement < 0) {
                        this.selectedElement = this.elements.$length() - 1;
                    }
                } while (this.elements.$get(selectedElement).action == null ||
                        this.elements.$get(selectedElement).isInactive());
                break;

            case RIGHT:
            case DOWN:
                do {
                    this.selectedElement++;
                    if (this.selectedElement >= this.elements.$length()) {
                        this.selectedElement = 0;
                    }
                } while (this.elements.$get(selectedElement).action == null ||
                        this.elements.$get(selectedElement).isInactive());
                break;

            default:
                res = false;
        }

        this.elements.$get(selectedElement).inFocus = true;

        return res;
    }

    public void draw(CanvasRenderingContext2D ctx) {
        for (int i = 0, e = this.elements.$length(); i < e; i++) {
            UiField el = this.elements.$get(i);

            if (!el.isInactive()) {
                el.draw(ctx);
            }
        }
    }
}
